### DISCUSSION TITLE
Getting 404 on the relative paths for my website after adding another repository with gh-pages

### DISCUSSION THREAD

**Discussion by aadictive** [community/community#50269] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

General

### Body

Hi All,

So I have my portfolio website hosted on GitHub pages. Here is the repository: [https://github.com/aadictive/aadictive.github.io](https://github.com/aadictive/aadictive.github.io)
The website is hosted on a custom domain, so it can be seen at [https://adityachaudhari.dev/](https://adityachaudhari.dev/)

Now when I click on the pages or tabs on my portfolio it takes me to the proper URL and displays the correct results. But when I refresh, I get a blank page or a 404 Not Found page. For example [https://adityachaudhari.dev/about](https://adityachaudhari.dev/about) or [https://adityachaudhari.dev/project](https://adityachaudhari.dev/project)

I recently added another repository called [test-app](https://github.com/aadictive/test-app) as I wish to add a few project repositories to my website. It opens up at [https://adityachaudhari.dev/test-app](https://adityachaudhari.dev/test-app)

I am not able to resolve the issue with 404 for my `/about`, `/project`, `/resume` links due to this new addition of repositories, it looks for repos with names 'about', 'projects', etc. Can someone help me resolve this issue?

I see a similar Issue for a month-old discussion: [here](https://github.com/orgs/community/discussions/47342#discussion-4856811
)

**Comment by Anmol-Baranwal** [comment]
Check your routing configuration: Review your website's routing configuration to ensure that it is correctly handling requests for the /about, /project, and /resume pages. Make sure that the routing rules are not conflicting with the new repository names.

Check your file names and paths: Make sure that the files for the /about, /project, and /resume pages exist and are located in the correct directories. Check the file names and paths to ensure that they match the routing configuration.

> you may have noticed, that we can visit those pages from home but cannot go it directly through the route URL. It may be a problem related to Tech Stack,

**Reply by Anmol-Baranwal** [reply]
@aadictive 
I just wanted to check in and see if you had any further questions or concerns about my answer on GitHub. If you have no further doubts or if your doubts have been resolved, would you mind marking my answer as accepted?

It would be greatly appreciated and it will also help other users who might be looking for a solution to a similar issue.

**Reply by aadictive** [reply]
It is still not working for me. Everything works correctly on my local machine as well as when deployed to Heroku. It is something to do with the GitHub pages I guess. /about is simply checking for another repository.
> Make sure that the routing rules are not conflicting with the new repository names.
^^ Thinking out loud what do you mean by this?? and how do I check that?

**Comment by aadictive** [comment]
I think I found the solution. For my project, all the relative links that have the repository name in them can route properly. I might have to do some additional settings with custom domain configs.
But essentially - When you have multiple pages or repositories hosted over Github Pages, make sure to follow this rule:
`{github-username}.github.io/{repository-name}/{page-name}`
For me if my main website has multiple pages like /about, /resume, etc. My URL looks like
`aadictive.github.io/aadictive.github.io/about` to be able to be correctly routed.
If I have another page in test-app-repository. The URL would be -
`aadictive.github.io/test-app/about`

Now if you find the repeating repository name in the first scenario like `aadictive.github.io/aadictive.github.io/` to be annoying, simply put a static page in the base repository and then keep portfolio or other projects in separate repos, so the URL would look a bit decent like - `aadictive.github.io/portfolio/about`

**Reply by aadictive** [reply]
Unfortunately, this did not work as well!

**Comment by aadictive** [comment, marked as answer]
To wrap this thread up!

I found the solution to my problem in the [comment](https://github.com/anitab-org/stem-diverse-tv-cms/issues/48#issuecomment-1120336354) in a [separate discussion](https://github.com/anitab-org/stem-diverse-tv-cms/issues/48). This solution is a temporary fix, but it worked like a charm!

FYI, the routing issue is a known issue in GitHub pages!


So first you will need to add a new page named 404.html to your repo.
```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>GitHub Pages Issue</title>
    <script type="text/javascript">  
 
      var pathSegmentsToKeep = 0;

      var l = window.location;
      l.replace(
        l.protocol + '//' + l.hostname + (l.port ? ':' + l.port : '') +
        l.pathname.split('/').slice(0, 1 + pathSegmentsToKeep).join('/') + '/?/' +
        l.pathname.slice(1).split('/').slice(pathSegmentsToKeep).join('/').replace(/&/g, '~and~') +
        (l.search ? '&' + l.search.slice(1).replace(/&/g, '~and~') : '') +
        l.hash
      );
    </script>
  </head>
  <body>
  </body>
</html>
```
If you are not using a custom domain on GitHub Pages set pathSegmentsToKeep = 0, otherwise set pathSegmentsToKeep = 1.

Copy this script and add to your index.html.
```
<script type="text/javascript">
  (function(l) {
    if (l.search[1] === '/' ) {
      var decoded = l.search.slice(1).split('&').map(function(s) { 
        return s.replace(/~and~/g, '&')
      }).join('?');
      window.history.replaceState(null, null,
          l.pathname.slice(0, -1) + decoded + l.hash
      );
    }
  }(window.location))
</script>
```
On your build folder need to have this file 404.html, and thats is it.

**Reply by cchanyu** [reply]
Wow, tysm for the update. this issue was one of the caveat of using gh-pages,

**Reply by Tarungarg98** [reply]
Thanks @aadictive, this helped a lot

**Reply by HunterCoders** [reply]
where to add this 404.html. I mean under which folder

**Reply by aadictive** [reply]
@HunterCoders You can put it under the `public` folder in you repo

**Reply by danielmelar** [reply]
it really works, thanks!  Even at Blazor wasm.
